### DISCUSSION TITLE
Events and listeners…

### DISCUSSION THREAD

**Discussion by gto406** [CppMicroServices/CppMicroServices#675] [Q&A] [answered]
Firstly, want to thank cpp-us folks for contributing this framework.  Really awesome!

I had a question re: general events (for a service) that go beyond service added/removed a-la the ServiceListenerHook mechanism.

Does cpp-us support a generalized event establish, listener/callback mechanism or is that delegated to the services (or other component)?

**Comment by achristoforides** [comment, marked as answer]
Hi @gto406 , thanks for your question! Apologies for not getting back to you sooner and thank you for the nice comments.

If I am understanding your question correctly, you are asking whether or not there is a general mechanism for publishing generic events and a general mechanism for listening/handling those events.

CppMicroServices currently does not have a generalized event publish/subscribe mechanism at this time. There is currently ongoing work for implementing the [Event Admin OSGi service spec](https://docs.osgi.org/specification/osgi.cmpn/7.0.0/service.event.html) but I cannot provide an estimate of when that work will be completed. It is on our list of things to implement though :)

@jeffdiclemente can also provide information regarding your question if I've missed anything.

Please let us know if you have any other questions or comments regarding this!
